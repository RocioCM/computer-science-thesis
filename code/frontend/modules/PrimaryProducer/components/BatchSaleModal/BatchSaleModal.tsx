import Modal, { ModalProps } from '@/common/components/Modal';

interface Props extends ModalProps {
  handleCancel: () => any;
  editingId: number | null;
  handleSuccess: () => void;
}

const BatchSaleModal: React.FC<Props> = ({
  editingId,
  handleCancel,
  handleSuccess,
  ...props
}) => {
  return (
    <Modal handleCancel={handleCancel} big skippable={false} {...props}>
      <h2 className="w-full bg-n0">Vender lote de botellas #{editingId}</h2>
    </Modal>
  );
};

export default BatchSaleModal;
