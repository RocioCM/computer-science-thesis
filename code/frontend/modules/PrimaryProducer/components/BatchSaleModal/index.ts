import BatchSaleModal from './BatchSaleModal';

export default BatchSaleModal;
