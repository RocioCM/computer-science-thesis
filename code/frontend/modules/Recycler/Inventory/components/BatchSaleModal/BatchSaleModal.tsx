import Modal, { ModalProps } from '@/common/components/Modal';
import RecyclerServices from '../../services';
import { SellRecyclingBatchPayload, Buyer } from '../../types';
import { toast } from 'react-toastify';
import Button from '@/common/components/Button';
import { SELL_FORM_INPUTS, SELL_FORM_STRUCT } from '../../constants';
import useForm from '@/common/hooks/useForm';

interface Props extends ModalProps {
  handleCancel: () => any;
  editingId: number | null;
  handleSuccess: () => void;
}

const BatchSaleModal: React.FC<Props> = ({
  editingId,
  handleCancel,
  handleSuccess,
  ...props
}) => {
  const { form, submitEnabled, formBuilder } = useForm(SELL_FORM_STRUCT);

  const getBuyers = async (query: string) => {
    const { data } = await RecyclerServices.searchBuyers(query);
    return (
      data?.map((buyer: Buyer) => ({
        label: `${buyer.userName}${buyer.userName ? ' - ' : ''}${buyer.email}`,
        value: buyer.firebaseUid,
      })) ?? []
    );
  };

  const handleSale = async () => {
    const payload: SellRecyclingBatchPayload = {
      batchId: editingId as number,
      buyerUid: form.buyerUid,
    };
    const { ok } = await RecyclerServices.sellRecyclingBatch(payload);
    if (ok) {
      toast.success(
        'Venta completada exitosamente. El lote reciclado ya está en propiedad del comprador'
      );
      handleCancel();
      handleSuccess();
    } else {
      toast.error(
        'No se pudo completar la venta. Por favor, inténtelo de nuevo más tarde'
      );
    }
  };

  return (
    <Modal handleCancel={handleCancel} {...props}>
      <h2 className="w-full bg-n0">Venta lote #{editingId}</h2>
      <div className="w-full flex flex-col gap-m my-xl">
        {formBuilder(SELL_FORM_INPUTS, {
          handleSearch: { buyerUid: getBuyers },
        })}
      </div>
      <Button
        handleClick={handleSale}
        disabled={!submitEnabled}
        width="full"
        label="Vender"
      />
    </Modal>
  );
};

export default BatchSaleModal;
