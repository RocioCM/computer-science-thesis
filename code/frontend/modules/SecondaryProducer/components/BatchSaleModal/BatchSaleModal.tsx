import Modal, { ModalProps } from '@/common/components/Modal';
import SecondaryProducerServices from '../../services';
import { SellProductBottlesPayload, Buyer } from '../../types';
import { toast } from 'react-toastify';
import Button from '@/common/components/Button';
import { SELL_FORM_INPUTS, SELL_FORM_STRUCT } from '../../constants';
import useForm from '@/common/hooks/useForm';

interface Props extends ModalProps {
  handleCancel: () => any;
  editingId: number | null;
  handleSuccess: () => void;
}

const BatchSaleModal: React.FC<Props> = ({
  editingId,
  handleCancel,
  handleSuccess,
  ...props
}) => {
  const { form, submitEnabled, formBuilder } = useForm(SELL_FORM_STRUCT);

  const getBuyers = async (query: string) => {
    const { data } = await SecondaryProducerServices.searchBuyers(query);
    return (
      data?.map((buyer: Buyer) => ({
        label: `${buyer.userName}${buyer.userName ? ' - ' : ''}${buyer.email}`,
        value: buyer.firebaseUid,
      })) ?? []
    );
  };

  const handleSale = async () => {
    const payload: SellProductBottlesPayload = {
      batchId: editingId as number,
      quantity: form.quantity,
      buyerUid: form.buyerUid,
    };
    const { ok } = await SecondaryProducerServices.sellBatch(payload);
    if (ok) {
      toast.success(
        'Venta completada exitosamente. Los productos seleccionados ya están en propiedad del comprador'
      );
      handleCancel();
      handleSuccess();
    } else {
      toast.error(
        'No se pudo completar la venta. Por favor, inténtelo de nuevo más tarde'
      );
    }
  };

  return (
    <Modal handleCancel={handleCancel} {...props}>
      <h2 className="w-full bg-n0">Venta lote #{editingId}</h2>
      <div className="w-full flex flex-col gap-m my-xl">
        {formBuilder(SELL_FORM_INPUTS, {
          handleSearch: { buyerUid: getBuyers },
        })}
      </div>
      <Button
        handleClick={handleSale}
        disabled={!submitEnabled}
        width="full"
        label="Vender"
      />
    </Modal>
  );
};

export default BatchSaleModal;
