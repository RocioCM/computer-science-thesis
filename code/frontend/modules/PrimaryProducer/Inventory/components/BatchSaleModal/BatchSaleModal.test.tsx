import { render, screen, fireEvent, act } from '@testing-library/react';
import BatchSaleModal from './BatchSaleModal';

describe('Primary Producer Inventory - BatchSaleModal', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    const modalRoot = document.createElement('div');
    modalRoot.setAttribute('id', 'modal-root');
    document.body.appendChild(modalRoot);
  });

  afterEach(() => {
    const modalRoot = document.getElementById('modal-root');
    if (modalRoot) {
      document.body.removeChild(modalRoot);
    }
  });

  it('Matches snapshot', () => {
    const { container } = render(
      <BatchSaleModal
        editingId={1}
        handleCancel={() => {}}
        handleSuccess={() => {}}
      />
    );
    expect(container).toMatchSnapshot();
  });
});
