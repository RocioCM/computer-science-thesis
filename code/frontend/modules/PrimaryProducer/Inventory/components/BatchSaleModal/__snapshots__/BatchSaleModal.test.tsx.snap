// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Primary Producer Inventory - BatchSaleModal Matches snapshot 1`] = `<div />`;
